### DISCUSSION TITLE
Can MemoryBlocks be compared

### DISCUSSION THREAD

**Discussion by Godlike-Productions** [RomanKubiak/ctrlr#189] [Q&A] [answered]
I think I remember reading somewhere that MemoryBlocks can't be compared.

so this is not valid
```
temp1B = MemoryBlock({0xf0)
temp2B = MemoryBlock()

temp2B = inputMB:getByte(0) -- assume inputMB's first byte is 0xf0

if temp1B == temp2B then
result = "true"
else
result = "false"
end --if
```

I think I remember reading that because MemoryBlocks are pointers, they will always give false even if the data is the same.  Is this correct.

**Comment by Ted-juh** [comment, marked as answer]
And what if you compare the bytes individually? 

If temp1B:getByte(1) == temp2B:getByte(1) then
result = true
else
result = false
end

Or a for loop to run over more than one byte at a time, might work as well.

**Reply by Godlike-Productions** [reply]
Yeah, I might need to do it that way,

Though Juce supports mb == mb, and they also have a function called matches, which returns true if the contents are the same, regardless of size.  I can't get them to work though. I think I'll just do getByte and do the comparison that way.

I'm trying to strip sysex headers.  The Jupiter has 2 different dumps that have different length headers, but that contain the same size data segment. I need to be able to pull from either and post to either, with a dynamically generated program number.

This is tricky.

**Reply by dnaldoog** [reply]
@DnaldeT's way or 

<pre>
local result="false"
if temp1B:toHexString(1) == temp2B:toHexString(1) then
     result="true"
end
</pre>

**Reply by Godlike-Productions** [reply]
That's even neater, thanks.

**Reply by Ted-juh** [reply]
I knew I could count on Dnaldoog to come up with a more suitable way to do it.

Does anyone know what these are for?
.def("compare", &BigInteger::compare)
.def("compareAbsolute",&BigInteger::compareAbsolute)

It's under the memory.cpp class for lua.

**Reply by Godlike-Productions** [reply]
I think the first compares 2 (probably 4 byte) ranges in memory to see if they match.  The second would compare the absolute values (ie -4 would match +4) doing the same thing.  As LUA is untyped you could probably use it to compare any memoryBlock range of the same length as BigInteger

**Reply by dnaldoog** [reply]
> I knew I could count on Dnaldoog to come up with a more suitable way to do it.
> 
> Does anyone know what these are for?
> .def("compare", &BigInteger::compare)
> .def("compareAbsolute",&BigInteger::compareAbsolute)
> 
> It's under the memory.cpp class for lua.

Got me curious! It returns -1,0 or +1 depending on whether one __BigInteger__ is less than, equal to or greater than another __BigInteger__. _compareAbsolute_ would appear to be for signed integers, so I just tested '_compare_'.


<pre>
changeValue = function(--[[ CtrlrModulator --]] mod --[[ number --]], value --[[ number --]], source)
    if source == 4 then
        local bigIntCompareResult = {[-1] = "less than", [0] = "equal to", "greater than"}
        mName = L(mod:getName())

        local cWith = "A"

        if mName == "a" then
            cWith = "B"
        end

        local bv = BigInteger(value)
        local bi = BigInteger(_G[cWith]:getModulatorValue())
        local c = bv:compare(bi)

        panel:getLabel("d"):setText(
            string.format(
                "A=%d and B=%d\nso A is %s B",
                A:getModulatorValue(),
                B:getModulatorValue(),
                bigIntCompareResult[c]
            )
        )
        panel:getLabel("d"):append(string.format("\nBigInteger()=%s", bi:toString(10, 0)))
        panel:getLabel("d"):append(string.format("\nBigInteger()=%s", bv:toString(10, 0)))
        panel:getLabel("d"):append(string.format("\nBigInteger:compare()=%d", c))
    end -- don't let this function run unless a uiSlider is moved!
end
</pre>

![bigintcompare](https://user-images.githubusercontent.com/76831683/109237678-693a3200-7815-11eb-9139-e4914bd3eb0d.png)

[BigInteger-Compare_1_0_0_2021-02-26_09-28.zip](https://github.com/RomanKubiak/ctrlr/files/6046608/BigInteger-Compare_1_0_0_2021-02-26_09-28.zip)

**Reply by Godlike-Productions** [reply]
Neither the toHexString(1) or getByte(0) methods worked on their own, and I still don't know why.

This is what I ended up doing to get it to work.

```
local sysexHeaderMB = MemoryBlock({0xf0})
local offset = 0
local returnMB = MemoryBlock()
local returnError = "error"

-- get the first 6 bytes
local tempMB = MemoryBlock()
local temp1B = MemoryBlock()
local maxIndex = tempMB:getSize() 
tempMB = MemoryBlock(inputMB:getRange(0x00,0x06))  --inputMB is the input in the function declaration. A Memory Block is parsed
temp1B = (inputMB:getByte(0))

    console (String("Header: "..sysexHeaderMB:toHexString(1)))
    console (String("Input: "..string.format("%02x", temp1B)))
    if sysexHeaderMB:toHexString(1) ~= string.format("%02x",temp1B) then 
    returnMB = inputMB
    returnError = "stripped"
    else 


        if tempMB:toHexString(1) == loadPatchHeaderMB:toHexString(1) then
        offset = 7
        -- strip
        inputMB:removeSection(0,offset)
        returnMB = inputMB
        returnError = "patch"
```

**Reply by dnaldoog** [reply]
First you are declaring temp1B as MemoryBlock, but a few lines down you are redeclaring it as a Lua variable which is interpreted as an integer.

Were you perhaps trying to do toHexString(1) on temp1B? That will not work.

**Comment by dnaldoog** [comment]
So if I understand correctly you are taking a _sysex_ message and substituting a header of potentially variable length to the message?

Didn't know about __removeSection()__. Thanks!

How about this?

<pre>
fuseMessage = function(--[[ CtrlrComponent --]] comp --[[ MouseEvent --]], event)
    local m, trunc = MemoryBlock(), MemoryBlock()
    local newHeader = L(panel:getLabel("in"):getText()) -- overwrite header
    local messageToAlter = L(panel:getLabel("inFull"):getText())  -- this is the original message
    m:loadFromHexString(newHeader)
    trunc:loadFromHexString(messageToAlter)

    trunc:removeSection(0, m:getSize())
    m:append(trunc)
    console("m => " .. m:toHexString(1))  -- new message to send! Yay!
end
</pre>

![stripHeaders](https://user-images.githubusercontent.com/76831683/109254226-f2149600-7834-11eb-8e0f-626a546e5322.png)

[strip headers_1_0_god-like-productions_2021-02-26_12-27.zip](https://github.com/RomanKubiak/ctrlr/files/6047337/strip.headers_1_0_god-like-productions_2021-02-26_12-27.zip)

**Reply by Godlike-Productions** [reply]
This is the function I have written.  I haven't tested it yet. Still sorting out a few bugs with moving patches around the library.

I add the patch headers when I export them as a .syx file (that can be loaded directly to the Jupiter 8), or during the patch load routine.

Basically I'll have a big long list of programs kept in the library.  The user can select which of the 128 memory slots they want to load it into for performance, or can just load it into the Jupiter's patch buffer.

```
function addPatchHeader(patchNumber,sysexData)

-- patchNumber is an integer between 0 and 127.  This is the patch location in the JP8 where this patch will be written
-- sysexData is the memoryBlock data block that has been stored in a patch.sysex (inside a patchbank).  This will have
-- had the headers stripped using stripSysexHeaders.
-- This method essentially reverses stripSysexHeaders.

-- We do this so that the data is portable (ie we can get it from any patch number and put it into any other patch number.
-- This method will not work for dumping to the buffer.  Put them on the panel (this can be done directly from the patch.sysex
-- with an offset of 0, and then dump from the panel, or we'll need another method. Actually we can probably make a case where
-- patchnumber == buffer

    local tempSysexMB = MemoryBlock()

    if patchNumber == "buffer" then
        local bufferHeader = {0xF0, 0x00, 0x00, 0x2F, 0x03, 0x02}
        tempSysexMB:createFromTable(bufferHeader)
        tempSysexMB:append(sysexData)
        return tempSysexMB
    elseif type(patchNumber) ~= "number" then -- it ain't "buffer" and it's not a number. We can't deal with this.
        return
    elseif (patchNumber >=0 and patchNumber <=127) then -- we should only be dealing with numbers now, but check range
        local bufferHeader = {0xF0, 0x00, 0x00, 0x2F, 0x03, 0x00, patchNumber}
        tempSysexMB:createFromTable(bufferHeader)
        tempSysexMB:append(sysexData)
        return tempSysexMB
    else -- we can't handle this number
        return
    end --if
end
```
edit: found a bug, so fixed it.

**Reply by dnaldoog** [reply]
<s>I don't understand how you are first comparing ==**patchNumber** to a string "buffer" then a string ~= "number" and then checking if **patchNumber** lies between -1 and 128? Where do "buffer" and "number" come from? 
If the if/else got past == "buffer" (because it is not "buffer") fine, but then if  **patchNumber**  is an integer then ~= "Number" will be true and the function will return/exit, so you can never get to the third condition __if (patchNumber >=0 and patchNumber <=127)__ as I see it!</s>

P.S.
Sorry I completely didn't see if type(patchNumber) != "number" - didn't see the type() function. Makes sense now! I was seeing if patchNumber~=number.

**Reply by Godlike-Productions** [reply]
They are parsed into the function.  LUA is typeless.

I call this function with

```
addPatchHeader("buffer",patchBank[5].sysex) -- for example or
addPatchHeader(15, patchBank[5].sysex)
```

Basically the function checks if it is returning a "buffer". If it is, return the buffer dump

If it isn't, it checks to make sure some other string hasn't been parsed, ensuring that a number is parsed.

Having determined that indeed a number , make sure it's within range, and return the relevant patch dump.

If it's out of range, it just returns.

I could have added error messages.

**Reply by dnaldoog** [reply]
Also if the header is always the same string, as it should be, then it's probably just easier to always store it statically with each 0-127 sysex message you are writing to. I did this on my C-PLG-150AN panel and the JD-990 panel. In the JD-990 panel I have a list of all 255 patches on the SR-JV80-04 card all stored as a table of sysex strings starting with F0...F7. In that particular case the checksum is also hard coded to the string. So 255 strings all with the same (9?) string header and EOX. It's a storage vs computing consideration. probably makes little difference in real speed. May even be faster.

**Reply by dnaldoog** [reply]
> They are parsed into the function. LUA is typeless.
> I call this function with
> 
> ```
> addPatchHeader("buffer",patchBank[5].sysex) -- for example or
> addPatchHeader(15, patchBank[5].sysex)
> ```

Sorry I completely didn't see __if type(patchNumber) ~= "number"__ - didn't see the type() function. Makes sense now! I was seeing __if patchNumber~=number__.  We see what we want to see!

**Reply by dnaldoog** [reply]
You could also do
<pre>
    local tempSysexMB = MemoryBlock()
       <i> local bufferHeader = {0xF0, 0x00, 0x00, 0x2F, 0x03, 0x02}</i>
    if patchNumber == "buffer" then
        tempSysexMB:createFromTable(bufferHeader)
        tempSysexMB:append(sysexData)
        return tempSysexMB
    elseif type(patchNumber) ~= "number" then -- it ain't "buffer" and it's not a number. We can't deal with this.
        return
    elseif (patchNumber >=0 and patchNumber <=127) then -- we should only be dealing with numbers now, but check range
        <b>table.insert(bufferHeader, patchNumber}</b>
        tempSysexMB:createFromTable(bufferHeader)
        tempSysexMB:append(sysexData)
        return tempSysexMB
    else -- we can't handle this number
        return
    end --if
end
</pre>

**Reply by Godlike-Productions** [reply]
For loading to the  Jupiter's Patch memory they are, but  there are also some other dump types that I haven't finished yet. 

Also I use the addPatchHeader function for dumping from the panel, which collects the onscreen controls from the plugin and assembles the entire dump dynamically.  

I guess I could just build a big table of all the possible headers, but I will be porting this code over to the Kurzweil panels which have tens of thousands of headers.

**Reply by Godlike-Productions** [reply]
Buffer header is different to patch header.  I just used the same variable name as I'm lazy when it came to copy paste.

The last byte of buffer header is 0x02, the last 2 bytes of a patch header are 0x00 and then the  program number, but yeah, I could have defined patchHeader and did the table insert trick.

**Reply by dnaldoog** [reply]
Okay, I see you've got good reasons for doing what you're doing!

I saw a cool thing I haven't used yet, but saw Atom using it in one of his panels.

<pre>
    local tempSysexMB = MemoryBlock()
        local bufferHeader = {0xF0, 0x00, 0x00, 0x2F, 0x03, 0x02}
    if patchNumber == "buffer" then
        tempSysexMB:createFromTable(bufferHeader)
        tempSysexMB:append(sysexData)
        return tempSysexMB
    elseif type(patchNumber) == "number" then -- it ain't "buffer" and it's not a number. We can't deal with this.
        <b>jlimit(0,127,patchNumber)</b> -- we should only be dealing with numbers now, but check range
        table.insert(bufferHeader, patchNumber}
        tempSysexMB:createFromTable(bufferHeader)
        tempSysexMB:append(sysexData)
        return tempSysexMB
    end --if
end
</pre>

https://docs.juce.com/master/group__juce__core-maths.html#gacf7cd562ab136d15d3bf5c3be047cf96

**Reply by Godlike-Productions** [reply]
Oh, that's nice.  I'll have a think about whether I should put that in.  Will need to think about the potential for out of range values and the impact on the data.

I definitely have another place I can use that thought (though probably jmin), that will replace a chunk of cludgy code I had to write to make sure that I can put new stuff into empty listboxes, without crashing everything.

Tables don't like being called with an index of -1 (if they start from 1)

**Comment by Godlike-Productions** [comment]
Weird bit of trivia for you.  I'm testing this from 350km away from my Jupiter 8.  Doing dumps across rtpMIDI.  It's interesting times we live in.

**Reply by dnaldoog** [reply]
> Weird bit of trivia for you. I'm testing this from 350km away from my Jupiter 8. Doing dumps across rtpMIDI. It's interesting times we live in.

Really? That is very interesting. Where are you and where is the Jupiter, if you don't mind saying? At least the Jupiter isn't on Jupiter!

I've see this [program](https://www.tobias-erichsen.de/software/rtpmidi.html) but never investigated.

**Reply by Godlike-Productions** [reply]
Jupiter is in Perth, Western Australia.  I'm on a mine site in Forrestania, Western Australia

**Reply by dnaldoog** [reply]
> Forrestania

Amazing! I am in Japan, - lived in Sydney for 20 years! We are in similar time zones.
